### PR TITLE
Add IDN URLs support to HTTP client

### PR DESCRIPTION
**Pull request by gsnewmark**
Currently it's not possible to make HTTP request to [IDN URLs](https://en.wikipedia.org/wiki/Internationalized_domain_name) using the Aleph because it relies on `java.net.URI` to [extract domain from the URL](https://github.com/ztellman/aleph/blob/9dcfc60321700ce271a41fc4765671f6f3bef5eb/src/aleph/http/client.clj#L58-L79) and `java.net.URI` doesn't really support IDNs: depending on constructor used it either silently ignores such domain names (`.getHost` returns `null`) or fails with exception (current Aleph behavior): 
```
user> (require '[aleph.http.client :as client])
nil
user> (aleph.http.client/req->domain {:url "https://名がドメイン.com"})
URISyntaxException Illegal character in hostname at index 8: https://名がドメイン.com  java.net.URI$Parser.fail (URI.java:2848)
user> *e
#error {
 :cause "Illegal character in hostname at index 8: https://名がドメイン.com"
 :via
 [{:type java.net.URISyntaxException
   :message "Illegal character in hostname at index 8: https://名がドメイン.com"
   :at [java.net.URI$Parser fail "URI.java" 2848]}]
 :trace
 [[java.net.URI$Parser fail "URI.java" 2848]
  [java.net.URI$Parser parseHostname "URI.java" 3387]
  [java.net.URI$Parser parseServer "URI.java" 3236]
  [java.net.URI$Parser parseAuthority "URI.java" 3155]
  [java.net.URI$Parser parseHierarchical "URI.java" 3097]
  [java.net.URI$Parser parse "URI.java" 3053]
  [java.net.URI <init> "URI.java" 673]
  [aleph.http.client$eval38639$req__GT_domain__38646 invoke "client.clj" 75]
  <REPL-related stacktrace...>
  ]}
```

I've added UTF to ASCII translation for hostname, but I'm not sure it's best possible solution, as with this approach there is one additional URL transformation for each request. What do you think about it?

On a side note, do you really need `java.net.URI` instance as a domain? If I don't miss anything it's used only as a [key for connection pool](https://github.com/ztellman/aleph/blob/9dcfc60321700ce271a41fc4765671f6f3bef5eb/src/aleph/http.clj#L213), wouldn't protocol+domain+port _string_ be enough for this purpose?

### CODE REVIEW

File: src/aleph/http/client.clj [hunk 1]
```diff
@@ -59,19 +60,20 @@
                (URI.
                  (name (or (:scheme req) :http))
                  nil
-                 (or (:host req) (:server-name req))
+                 (when-some [host (or (:host req) (:server-name req))]
```
Comment (danielcompton) [inline src/aleph/http/client.clj]: A (subjective) matter of taste, but `some->` might be easier to read here. Will let the author of Elements of Clojure make the call though :)

File: test/aleph/http/client_test.clj [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                                   :port 80})]
+      (is (= "https" (.getScheme uri)))
+      (is (= "domainname.com" (.getHost uri)))
+      (is (= 80 (.getPort uri)))))
+  (testing "IDN domain in URL is translated to ASCII."
+    (let [uri (client/req->domain {:url "https://名がドメイン.com:80"})]
+      (is (= "https" (.getScheme uri)))
+      (is (= "xn--v8jxj3d1dzdz08w.com" (.getHost uri)))
+      (is (= 80 (.getPort uri)))))
+  (testing "IDN domain from request host is translated to ASCII."
```
Comment (danielcompton) [inline test/aleph/http/client_test.clj]: Could be good to add a test for an [Internationalized TLD](https://en.wikipedia.org/wiki/Internationalized_country_code_top-level_domain) too.

File: src/aleph/http/client.clj (hunk 1 above)
Comment (gsnewmark) [inline src/aleph/http/client.clj]: agree, `some->` would definitely be more readable, didn't think about it

**Comment by ztellman** [comment]
Netty expects a URI, so that's why I use it.  This looks good to me.

**Comment by gsnewmark** [comment]
@ztellman thanks!

**Comment by ztellman** [comment]
Is this blocking you? I can cut an alpha release, if so.

**Comment by gsnewmark** [comment]
@ztellman not really, it affects only a small part of code and we have a workaround in place (basically transforming URL before passing it to Aleph). So no need for a new alpha. 

By the way, are there any plans for a new "stable" release or there are not enough changes yet?

**Comment by ztellman** [comment]
There are a few changes I'm contemplating in Manifold w.r.t. to upstream propagation of closing, which are delaying the release of the whole suite of libraries.  I'll have a clearer answer for that in a week.

**Pull request merged**